Architecture overview (#161)
trilitech/tezos-developer-docs: docs/architecture.mdx

Reviewer: I apologize for my quite radical recommendation, but although there are some nice ideas in this page, I think the whole page must be restarted over, because there are multiple critical issues with the current discourse, e.g.:
- the notions of "nodes" differentiated in "baking nodes" and "accuser nodes" is quite imprecise (e.g. there are nodes that neither bake nor accuse, just to consensus, and optionally serve RPCs) and leads to confusions (e.g. not only baking nodes maintain the context, but also the consensus-only nodes)
  * it's important for users to know that they can participate in the consensus without being bakers
  * although it seemed a good idea to introduce a "baking/non-baking node" abstraction, we should be very cautious not to conflict with the notions of node, baker, and accuser, that are commonly used today, by users and bakers alike. Finally, I think we'd better stick to the current notions, to avoid the risk of a big terminology chaos
- the list of components is debatable in my opinion: is blockchain data a component? At least, it's very different from the other, which are executables (= processes)
  * talking about several Tezos networks at the same time adds unnecessary complexity (the first picture could only concern one network, while some phrase could just link to the notion of main and test networks)
- the discussion about blockchain trees vs single chain also adds complexity and should be abstracted away using the notion of consensus between nodes
- the network is already quite complicated to understand, so perhaps the L1 network should be first presented alone, and then only including rollups in a separate picture (with perhaps a simplified cloud for the L1)
- the discussion of mempools is misplaced in the section about clients, as it's something internal to the nodes
- etc.
Therefore, instead of proposing fixes to this version, I propose to start anew from the proposal in <https://gitlab.com/nomadic-labs/tezos/-/blob/nic@recenter-octez-proto-doc/docs/global/architecture.rst>, and adapt it to fit some of the ideas in the current version of your page. In a first time, we can draft this back within the gdoc, before updating this PR.

Reviewer: I added a concrete proposal in the gdoc.

Author: Thank you for the review — I'll take a hard look at the structure and think about where to take this info.

---
title: Architecture
authors: Tim McMackin
last_update:
  date: 20 November 2023
---

import LucidDiagram from '@site/src/components/LucidDiagram';

The Tezos blockchain is composed of many Tezos nodes running around the world, complemented by other running daemons, such as bakers and accusers.
These processes collaborate with the overall goal of maintaining the blockchain data in a decentralized manner.

The Tezos nodes are the most important piece in this architecture because they maintain the system and the blockchain data.
Users interact with nodes through many different clients, including command-line clients, wallets, and web applications.

This diagram shows a high-level view of the Tezos system:

> <LucidDiagram width="640px" height="480px" src="https://lucid.app/documents/embedded/d778aa2a-ad0a-4324-b235-ed3b35742c58" id="Uz-K28y3R8gJ" />

Reviewer 2: The lucid diagram here doesn't load for me.

## The blockchain data

Although people often use the word "blockchain" to mean the entire system, strictly speaking, a Tezos blockchain is a series of blocks of data, each connected to the previous block in the chain, beginning with the genesis block.
The blockchain data is maintained by a network of Tezos nodes.
Nodes reach consensus on the next block before adding it to the chain.

> There are several Tezos blockchains.

Reviewer: Make the connection with the usual term "network" that you use right after:
```suggestion
There are several Tezos blockchains, commonly called Tezos networks.
```

- It implements a client/server architecture, to allow composition with many other tools in a safe way.

The Octez suite, which is an implementation of the Tezos node and other executables, instantiates these principles in the [Octez software architecture](https://tezos.gitlab.io/shell/the_big_picture.html).

These two principles are detailed next.

## Shell & Protocol

The software that runs Tezos nodes is split into two main parts:

- The shell, which handles low-level functions like data storage and peer-to-peer network communication
- The economic protocol that interprets the block operations (e.g., transactions)

The relationship between the shell and the protocol is like the relationship between an operating system and an application.
The operating system stays stable while the application can update itself.
In this way, Tezos can update how it works (its protocol) without requiring nodes to accept major changes to the software that they run (the shell).
For example, nodes can update to a new protocol version without restarting the shell.

### The Tezos self-amending protocol

Unlike many other blockchains, Tezos is self-amending.

> Its nodes can update the protocol that controls the possible operations and how they are processed via an online governance process.

Reviewer: If you don't like the comma before "via", let's separate in two sentences, to avoid the wrong reading "processed via an online governance process":
```suggestion
Its nodes can update the protocol that controls the possible operations and how they are processed; updates are performed via an online governance process.
```

These updates allow Tezos to adapt to new technologies and respond to user needs.
For example, protocol upgrades have added new features like Smart Rollups and have reduced the amount of time between blocks.

This protocol is responsible for interpreting the operations in each block.
It also provides the logic that identifies erroneous blocks.

> Updates to the protocol can change this logic through a voting process.

Reviewer: Now that the governance has been already announced, let's rephrase to avoid repetition:
```suggestion
The online governance managing protocol updates involves a voting process.
```

> Updates to the protocol can change this logic through a voting process.

Reviewer: Here you could mention Voting operations if you included them in the picture, e.g:
```suggestion
Updates to the protocol can change this logic through a voting process, using dedicated voting operations such as protocol proposals or protocol upvotes.
```

- Storage functionality that lets nodes store blocks, operations, and the current state of the chain
- A synchronization heuristic that starts nodes and keeps them in sync with the network
- A validator that checks that blocks are valid with help from the rules in the economic protocol

In particular, the validator is responsible for resolving the available blocks into a single linear sequence of blocks.
It chooses between the various blocks that baking nodes create, uses the protocol to verify and score them, and selects the tree head with the highest score.
Then it uses that linear chain in all of its work with the protocol, so the protocol is never aware of multiple branches.

## Tezos clients and servers

In addition to the functions of the protocol and shell, a Tezos node also acts as a server to respond to queries and requests from clients.
A client can query the chain’s state and can inject blocks and operations into a node.

Tezos uses this client-server architecture for these main reasons:

- It improves security by separating the node, which is exposed to the internet, from the baker, which has access to the client keys.

> The node and the baker can be on different computers, which allows the node to manage communication and shield bakers from network attacks.

Reviewer: ```suggestion
The node and the baker can be on different computers, which allows the node to manage communication and shields bakers from network attacks.
```

The node accepts calls from clients through its RPC interface.
It has control over which clients to accept calls from, which calls to accept, or whether to accept RPC calls at all.
For more information on the RPC interface, see [The RPC protocol](./architecture/rpc).

### The baker daemon

The baker daemon is an Octez program that is responsible for creating and proposing new blocks based on the operations proposed by different clients.
The baker daemon is associated with an account, which means that it has access to the account’s private key, which it uses to sign blocks and operations.

### The accuser daemon

The accuser daemon is an Octez program that monitors new blocks and looks for problems, such as when bakers try to add more than one block at a time.
When it finds a problem, it submits a denunciation to other nodes to refuse the new blocks and punish the offending node.

### The Octez client

The Octez client is a command-line tool that developers can use for many Tezos-related tasks, including:

- Deploying, calling, testing, and interacting with contracts
- Deploying and interacting with Smart Rollups
- Working with accounts

> - Getting contract balances and data

Author: Removed.

- Calling RPC endpoints directly

> - Running sapling transactions

Reviewer: FYI, in the Octez doc, we always spell Sapling with a capital, but up to you to choose your convention (be consistent, though).

Author: I see that opentezos capitalizes Sapling as well -- I'll make that update.

> 

Author: Added item "Setting up baking operations for testing contracts"
